packages/fetch/src/modifiers.ts

```diff
@@ -0,0 +1,204 @@
+ * retry, but never more than 30000.
+ * @param verify a function that receives the response and verifies if it is valid;
+ * default checks if response.ok is true.
+ */
+export const withRetry: RequestModifier =
+  <Result extends unknown, FetcherArgs extends any[]>(
+    retries: number,
+    wait: number | ((attempt: number) => number) = defaultWait,
+    verify = (response?: Response) => response?.ok
+  ) =>
+  requestContext => {
+    const waitForAttempt = (attempt: number) =>
+      new Promise<void>(resolve =>
+        setTimeout(resolve, typeof wait === "number" ? wait : wait(attempt))
+      );
+    wrapFetcher<Result, FetcherArgs>(
+      requestContext as unknown as RequestContext<Result, FetcherArgs>,
+      originalFetcher => (requestData: FetcherArgs, info: ResourceFetcherInfo<Result>) => {
+        const wrappedFetcher = (attempt: number): Promise<Result> =>
+          originalFetcher(requestData, info)
+            .then(data =>
+              !verify(requestContext.response) && attempt <= retries
+                ? waitForAttempt(attempt).then(() => wrappedFetcher(attempt + 1))
+                : data
+            )
+            .catch(err =>
+              attempt > retries
+                ? Promise.reject(err)
+                : waitForAttempt(attempt).then(() => wrappedFetcher(attempt + 1))
+            );
+        return wrappedFetcher(0);
+      }
+    );
+    requestContext.wrapResource();
+  };
+
+export type RefetchEventOptions<Result extends unknown, FetcherArgs extends any[]> = {
+  on?: (keyof WindowEventMap)[];
+  filter?: (requestData: FetcherArgs, data: Result | undefined, ev: Event) => boolean;
+};
+
+export const withRefetchEvent: RequestModifier =
+  <Result extends unknown, FetcherArgs extends any[]>(
+    options: RefetchEventOptions<Result, FetcherArgs> = {}
+  ) =>
+  (requestContext: RequestContext<Result, FetcherArgs>) => {
+    const lastRequestRef: { current: [requestData: FetcherArgs, data?: Result] | undefined } = {
+      current: undefined
+    };
+    wrapFetcher<Result, FetcherArgs>(requestContext, originalFetcher => (...args) => {
+      lastRequestRef.current = [args as any, undefined];
+      return originalFetcher(...args).then(data => {
+        lastRequestRef.current = [args as any, data as Result];
+        return data;
+      });
+    });
+    requestContext.wrapResource();
+    const events: string[] = options.on || ["visibilitychange"];
+    const filter = options.filter || (() => true);
+    const handler = (ev: Event) => {
+      if (
+        lastRequestRef.current &&
+        !(ev.type === "visibilitychange" && document.visibilityState !== "visible") &&
+        filter(...lastRequestRef.current, ev)
+      ) {
+        requestContext.resource?.[1].refetch();
+      }
+    };
+    events.forEach(name => window.addEventListener(name, handler));
+    getOwner() &&
+      onCleanup(() => events.forEach(name => window.removeEventListener(name, handler)));
```

Comment:
+1 on the `window.` thing. Also, this should probably be wrapped with `if (!isServer) {}` anyway.
You are accessing `document.visibilityState` higher up as well.

Reply:
It won't work on the server anyways, so I'll just replace it with a shim there.